### PR TITLE
Persist and reload message counters on the controller

### PR DESCRIPTION
**Pull request by sagar-apple**
#### Problem
The controller never persists message counters used for the PASE session but allows persisting the session in storage. This means that each time a connection is stored away and reused later, the counter values are lost. 
Resuming these PASE sessions from storage causes the counters to go out of sync and messages get dropped. 

I'm seeing this happen every time while trying to provision an M5Stack via the chip-tool cli

#### Change overview
- Persist the session message counters every time a CHIPDevice is released/destroyed. 
- Reload the message counters when a session is reestablished. 
- Minor unrelated code cleanup in objc 

#### Testing
How was this tested? (at least one bullet point required)
* If new unit tests are not added, why not?
> While I did get some coverage from the Darwin cluster tests, I'm not entirely sure how to unit test this specific change. We're missing some test infrastructure here that prevents me from adding a test easily. I need mock persistent storage + the ability to create and destroy CHIPDevices + connections to ensure they're persisting values correctly.
* If manually tested, what platforms controller and device platforms were manually tested, and how?
> Tested ble-pairing via the `chip-tool` cli on a mac and  `CHIPTool` on iOS with the all-clusters-app on a M5Stack. Verified that while pairing, message counter errors are no longer seen and that as we transition from pairing -> network provisioning the messages coming in are being sent with the right message counter values.

### CODE REVIEW

File: src/controller/CHIPDevice.cpp [hunk 1]
```diff
@@ -470,5 +496,29 @@ void Device::AddReportHandler(EndpointId endpoint, ClusterId cluster, AttributeI
...
+        // Ensure that any exchange contexts we have open get closed now,
+        // because we don't want them to call back in to us after this
+        // point.
+        mExchangeMgr->CloseAllContextsForDelegate(this);
+    }
+
+    if (mStorageDelegate)
+    {
+        // Store the current device in persistent storage so we have the latest
+        // message counters available next time.
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.cpp:512]: So we need to do this because:

1. When we deserialize we will deserialize the CASE key information.
2. We don't currently do CASE resumption.

and therefore the fact that there is serialization/deserialization happening here is an implementation detail; it's all one CASE session as far as the spec is concerned, and hence we need message counter continuity.

Is that a correct description of the situation?  If so, it would be good to have the comments (probably at the spot where we deserialize) make this clear.  If not, whatever the reason we need to do this might be, it should be documented in the deserialization code.

File: src/messaging/ExchangeMgr.cpp [hunk 2]
```diff
@@ -198,8 +198,8 @@ void ExchangeManager::OnMessageReceived(const PacketHeader & packetHeader, const
     UnsolicitedMessageHandler * matchingUMH = nullptr;
     bool sendAckAndCloseExchange            = false;
 
-    ChipLogProgress(ExchangeManager, "Received message of type %d and protocolId %d", payloadHeader.GetMessageType(),
-                    payloadHeader.GetProtocolID());
+    ChipLogProgress(ExchangeManager, "Received message of type %d and protocolId %d on exchange %d", payloadHeader.GetMessageType(),
+                    payloadHeader.GetProtocolID(), payloadHeader.GetExchangeID());
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeMgr.cpp:202]: The exchange id is not really meaningful unless we're the initiator (i.e. the I flag is _not_ set here) or unless we also log the source node information for the message....  This is probably OK for now, but maybe a followup to have better logging here?

File: src/transport/MessageCounter.h [hunk 3]
```diff
@@ -71,6 +72,12 @@ class GlobalUnencryptedMessageCounter : public MessageCounter
         ++value;
         return CHIP_NO_ERROR;
     }
+    CHIP_ERROR SetCounter(uint32_t count) override
```
Comment (bzbarsky-apple) [inline src/transport/MessageCounter.h:75]: When would we ever want to set this value for this counter type?

File: src/transport/MessageCounter.h [hunk 4]
```diff
@@ -98,13 +106,21 @@ class GlobalEncryptedMessageCounter : public MessageCounter
     {
         FakePersistedCounter() : value(0) {}
         CHIP_ERROR Init(chip::Platform::PersistedStorage::Key aId, uint32_t aEpoch) { return CHIP_NO_ERROR; }
-
+        void Reset()
+        { /* null op */
+        }
```
Comment (bzbarsky-apple) [inline src/transport/MessageCounter.h:111]: Why is this needed?

File: src/transport/MessageCounter.h [hunk 5]
```diff
@@ -98,13 +106,21 @@ class GlobalEncryptedMessageCounter : public MessageCounter
...
         {
             ++value;
             return CHIP_NO_ERROR;
         }
+        CHIP_ERROR SetCounter(uint32_t count)
+        {
+            Reset();
+            value = count;
+            return CHIP_NO_ERROR;
+        }
```
Comment (bzbarsky-apple) [inline src/transport/MessageCounter.h:123]: Why is this needed?  Seems to be unused....

File: src/transport/MessageCounter.h [hunk 6]
```diff
@@ -127,6 +143,12 @@ class LocalSessionMessageCounter : public MessageCounter
         ++value;
         return CHIP_NO_ERROR;
     }
+    CHIP_ERROR SetCounter(uint32_t count) override
```
Comment (bzbarsky-apple) [inline src/transport/MessageCounter.h:146]: Might want to document why we'd want to do this....

File: src/transport/MessageCounter.h (hunk 3 above)
Comment (sagar-apple) [reply src/transport/MessageCounter.h:75]: This is the counter used for local message counting (as opposed to the `PeerMessageCounter`). This counter has no mechanism to set its value to a specified count, so instead of reset+advance in a for loop, I added a SetCounter API to just reset+set it to the desired value. 

CHIPDevice is doing this when a session is established.

File: src/messaging/ExchangeMgr.cpp (hunk 2 above)
Comment (sagar-apple) [reply src/messaging/ExchangeMgr.cpp:202]: Yeah this isn't very interesting by itself. I was using it with debug logs in send to figure out the lifetimes of exchanges while debugging more issues that lie beyond the one this PR is fixing.

File: src/transport/MessageCounter.h (hunk 3 above)
Comment (bzbarsky-apple) [reply src/transport/MessageCounter.h:75]: Right, but my understanding is that in the spec this counter is meant to just be set to a random value on boot...  But given that we then do value enforcement for it, that seems like a pretty serious spec bug!  @turon @tcarmelveilleux how is this part supposed to work?  The random value can randomly fall into a range the other side will consider a duplicate!

For now, I guess we keep this; it's almost certain that _this_ is why we were seeing handshake failures randomly: it depended on where the counter value fell.

Comment (bzbarsky-apple) [reply src/transport/MessageCounter.h:75]: I filed https://github.com/project-chip/connectedhomeip/issues/7108 to track part of this....

File: src/controller/CHIPDevice.cpp [hunk 7]
```diff
@@ -470,5 +496,29 @@ void Device::AddReportHandler(EndpointId endpoint, ClusterId cluster, AttributeI
...
+        Transport::PeerConnectionState * connectionState = mSessionManager->GetPeerConnectionState(mSecureSession);
+        VerifyOrReturn(connectionState != nullptr);
+        SerializedDevice serialized;
+        Serialize(serialized);
+        // TODO: no need to base-64 the serialized values AGAIN
```
Comment (msandstedt) [inline src/controller/CHIPDevice.cpp:517]: I thought our persistent data store interfaces were opaque binary now.  Why the need to base-64 encode at all?

File: src/controller/CHIPDevice.cpp [hunk 8]
```diff
@@ -470,5 +496,29 @@ void Device::AddReportHandler(EndpointId endpoint, ClusterId cluster, AttributeI
     mCallbacksMgr.AddReportCallback(mDeviceId, endpoint, cluster, attribute, onReportCallback);
 }
 
+Device::~Device()
```
Comment (msandstedt) [inline src/controller/CHIPDevice.cpp:499]: I would request that this device self-persistence be refactored as a method on the device object that the device simply calls in its destructor.

Right now, we have device persistence code duplicated here and in CHIPDeviceController.cpp.

File: src/lib/support/BUILD.gn [hunk 9]
```diff
@@ -77,6 +77,7 @@ static_library("support") {
     "LifetimePersistedCounter.h",
     "PersistedCounter.cpp",
     "PersistedCounter.h",
+    "PersistentStorageUtils.h",
```
Comment (andy31415) [inline src/lib/support/BUILD.gn:80]: Please use names that are not utils/helpers/support/misc.

Maybe `PersistenceDeviceOperationMacros.h`

File: src/lib/support/PersistentStorageUtils.h [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+namespace chip {
+
+constexpr const char kPairedDeviceListKeyPrefix[] = "ListPairedDevices";
+constexpr const char kPairedDeviceKeyPrefix[]     = "PairedDevice";
+constexpr const char kNextAvailableKeyID[]        = "StartKeyID";
+
+// This macro generates a key for storage using a node ID and a key prefix, and performs the given action
+// on that key.
+#define PERSISTENT_KEY_OP(node, keyPrefix, key, action)                                                                            \
```
Comment (andy31415) [inline src/lib/support/PersistentStorageUtils.h:36]: could we replace this macro with a method or class instead?

Macros are harder to follow in code.

Comment (sagar-apple) [reply src/lib/support/PersistentStorageUtils.h:36]: I can file an issue for that after, but don't you think that's outside the scope of this change?

File: src/controller/CHIPDevice.cpp (hunk 7 above)
Comment (sagar-apple) [reply src/controller/CHIPDevice.cpp:517]: This TODO (which came with the copy-paste) is just talking about how the storage delegate ends up base64 encoding everything that gets passed in. In this case the device is already serialized to base64 and when we send it off to storage, it gets encoded a second time.

Comment (msandstedt) [reply src/controller/CHIPDevice.cpp:517]: Ah, I see.

So that's platform-specific behavior though, isn't it?  If so, I don't think this comment actually makes sense.

It's not a big deal though.

**Comment by turon** [comment]
The specification calls for only persisting the global message counters used for group messages.  Session counters are only meant to persist for the life of a session, and sessions are to be reestablished from scratch if there is a sync issue or reboot.

File: src/controller/CHIPDevice.cpp (hunk 1 above)
Comment (sagar-apple) [reply src/controller/CHIPDevice.cpp:527]: The CHIPDevice currently has PASE sessions keys which are used for provisioning. The current controller design allows the CHIPDevice to be recreated in the middle of the provisioning flow. This is different from things like OpCreds which are provisioned internally without application input. 

Just as we're ready to provision, the Controller saves the CHIPDevice to storage and notifies the application. Which then recreates the CHIPDevice for network provisioning. 

When it's recreated, the same PASE keys are used to communicate with the device. So for now we've got to make sure we restore the entire state correctly for the PASE session to work, and that's why we need to persist the message counters as well..

File: src/controller/CHIPDevice.cpp [hunk 11]
```diff
@@ -164,6 +169,14 @@ CHIP_ERROR Device::Serialize(SerializedDevice & output)
     serializable.mDevicePort = Encoding::LittleEndian::HostSwap16(mDeviceAddress.GetPort());
     serializable.mAdminId    = Encoding::LittleEndian::HostSwap16(mAdminId);
 
+    Transport::PeerConnectionState * connectionState = mSessionManager->GetPeerConnectionState(mSecureSession);
+    VerifyOrExit(connectionState != nullptr, error = CHIP_ERROR_INCORRECT_STATE);
+    localMessageCounter = connectionState->GetSessionMessageCounter().GetLocalMessageCounter().Value();
+    peerMessageCounter  = connectionState->GetSessionMessageCounter().GetPeerMessageCounter().GetCounter();
```
Comment (kghost) [inline src/controller/CHIPDevice.cpp:175]: peerMessageCounter shouldn't be persistent, it will be negotiated during CASE resumption.

If it is needed temporary to fix CI problem, please add a TODO issue.

File: src/transport/MessageCounter.h (hunk 3 above)
Comment (kghost) [reply src/transport/MessageCounter.h:75]: Let's continue the discussion in #7108

File: src/controller/CHIPDevice.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply src/controller/CHIPDevice.cpp:527]: OK, so the same thing with PASE instead of CASE: doing a serialize/deserialize in what is conceptually (in spec terms) the same PASE session, because we use the same keys and don't redo the handshake.

File: src/transport/MessageCounter.h (hunk 3 above)
Comment (bzbarsky-apple) [reply src/transport/MessageCounter.h:75]: To be clear, this PR and the problem it's trying to solve is not related to #7108 per se.  This PR is about the fact that our code has a serialize/deserialize in the middle of what is conceptually a single PASE session from the spec's point of view.  And if we do that, we need to keep counters across that operation, since it's the same session.

**Review by msandstedt — APPROVED** [review]

File: src/controller/CHIPDevice.cpp (hunk 11 above)
Comment (sagar-apple) [reply src/controller/CHIPDevice.cpp:175]: https://github.com/project-chip/connectedhomeip/pull/7106/files/71468cce3a47939418b33319ad549fd64ede5c9f#r639366772
I think this comment explains why this is needed right now.

File: src/controller/CHIPDevice.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply src/controller/CHIPDevice.cpp:527]: @sagar-apple Can you please add the code comments for this?

File: src/controller/CHIPDevice.cpp [hunk 12]
```diff
@@ -258,10 +273,38 @@ CHIP_ERROR Device::Deserialize(const SerializedDevice & input)
...
+}
+
 void Device::OnNewConnection(SecureSessionHandle session)
 {
     mState         = ConnectionState::SecureConnected;
     mSecureSession = session;
+
+    // Reset the message counters here because this is the first time we get a handle to the secure session.
+    // Since CHIPDevices can be serialized/deserialized in the middle of what is conceptually a single PASE session
+    // we need to restore the session counters along with the the session information.
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.cpp:298]:
```suggestion
    // we need to restore the session counters along with the session information.
```

**Review by bzbarsky-apple — APPROVED** [review]

File: src/lib/support/PersistentStorageUtils.h (hunk 10 above)
Comment (andy31415) [reply src/lib/support/PersistentStorageUtils.h]: agree, just moving is ok, not for this PR.

File: src/controller/CHIPDevice.h [hunk 13]
```diff
@@ -494,6 +499,8 @@ typedef struct SerializableDevice
     uint8_t mDeviceTransport;
     uint8_t mDeviceProvisioningComplete;
     uint8_t mInterfaceName[kMaxInterfaceName];
+    uint32_t mLocalMessageCounter; /* This field is serialized in LittleEndian byte order */
```
Comment (andy31415) [inline src/controller/CHIPDevice.h:502]: should these comments be part of the struct? I believe that this should be part of the serialize/deserialize methods like "serializes 16-bit and larger values as little endian".

File: src/controller/CHIPDeviceController.cpp [hunk 14]
```diff
@@ -485,19 +468,9 @@ CHIP_ERROR DeviceController::UpdateDevice(Device * device, uint64_t fabricId)
 
 void DeviceController::PersistDevice(Device * device)
 {
-    // mStorageDelegate would not be null for a typical pairing scenario, as Pair()
-    // requires a valid storage delegate. However, test pairing usecase, that's used
-    // mainly by test applications, do not require a storage delegate. This is to
-    // reduce overheads on these tests.
-    // Let's make sure the delegate object is available before calling into it.
-    if (mStorageDelegate != nullptr && mState == State::Initialized)
+    if (mState == State::Initialized)
```
Comment (andy31415) [inline src/controller/CHIPDeviceController.cpp:471]: please add a log if not initialized, to not silently fail persistence.

Ideally Persist would return a CHIP_ERROR. If I call "persistdevice" I do not expect it to potentially silently fail.

File: src/controller/CHIPDevice.cpp [hunk 15]
```diff
@@ -258,10 +273,38 @@ CHIP_ERROR Device::Deserialize(const SerializedDevice & input)
...
 
+CHIP_ERROR Device::Persist()
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    if (mStorageDelegate != nullptr)
+    {
+        SerializedDevice serialized;
+        Serialize(serialized);
+
+        // TODO: no need to base-64 the serialized values AGAIN
```
[the remark this answers is not in the document]
Comment (sagar-apple) [reply src/controller/CHIPDevice.cpp:284]: I think we are still base-64 encoding it twice right? Once because the device supports serializing to a `SerializedDevice` and then again because the Darwin storage delegate converts everything to base-64

File: src/controller/CHIPDeviceController.cpp (hunk 14 above)
Comment (sagar-apple) [reply src/controller/CHIPDeviceController.cpp:471]: So the Device does return an error but the controller code entirely ignores missing storage delegates (because it's missing only in tests). 

I've updated it to log errors from Persist() on the CHIPDevice and I've added a log if the state is not initialized

File: src/controller/CHIPDevice.h [hunk 16]
```diff
@@ -226,6 +218,13 @@ class DLL_EXPORT Device : public Messaging::ExchangeDelegate, public SessionEsta
      **/
     CHIP_ERROR Deserialize(const SerializedDevice & input);
 
+    /**
+     * @brief Serialize and store the Device in persistent storage
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
```
[the remark this answers is not in the document]
Comment (sagar-apple) [reply src/controller/CHIPDevice.h:224]: I thought Doxygen requires us to describe return values? I'll see if I can describe it in any more detail.

File: src/controller/CHIPDevice.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (andy31415) [reply src/controller/CHIPDevice.cpp:284]: it was non-obvious to me because 'inner' is not type-defined.

if it would be inner.base64 that would be understandable. However now it is inner and sizeof(inner) so everything sure looks binary to me.

Consider moving inside the serialize.inner reading/writing since this is were we say "this is binary, no need for base64".

File: src/controller/CHIPDevice.cpp [hunk 17]
```diff
@@ -258,10 +273,43 @@ CHIP_ERROR Device::Deserialize(const SerializedDevice & input)
...
+    // Reset the message counters here because this is the first time we get a handle to the secure session.
+    // Since CHIPDevices can be serialized/deserialized in the middle of what is conceptually a single PASE session
+    // we need to restore the session counters along with the session information.
+    Transport::PeerConnectionState * connectionState = mSessionManager->GetPeerConnectionState(mSecureSession);
+    VerifyOrReturn(connectionState != nullptr);
+    MessageCounter & localCounter = connectionState->GetSessionMessageCounter().GetLocalMessageCounter();
+    if (localCounter.SetCounter(mLocalMessageCounter))
+    {
+        ChipLogError(Controller, "Unable to restore local counter to %d", mLocalMessageCounter);
+    };
```
Comment (andy31415) [inline src/controller/CHIPDevice.cpp:310]: nit: unnecessary ;

**Review by andy31415 — APPROVED** [review]
